src/librustc_ast_lowering/lib.rs

```diff
@@ -2281,6 +2281,7 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
             }
             StmtKind::Expr(ref e) => hir::StmtKind::Expr(self.lower_expr(e)),
             StmtKind::Semi(ref e) => hir::StmtKind::Semi(self.lower_expr(e)),
+            StmtKind::Empty => hir::StmtKind::Semi(self.expr_unit(s.span)),
```

Comment:
Can it result in zero statements in HIR?

Reply:
Seems like it can -- at least I can't think of a situation where it would affect type checking, so I've tweaked lowering accordingly. The only case I can think of where it matters is:
```rust
fn foo() {
    return;
    ; //~ warning: unreachable statement
}
```